File: cuecard-mobile/ios/CueCard/CueCard/Models/RemoteConfig.swift
import Foundation

/// The payload behind `GET /v2/config` on the mobile worker.
///
/// Decoding is deliberately forgiving. Every flag defaults to its permissive
/// value, so a build that can't reach the worker behaves exactly as shipped, and
/// anything a newer worker adds that this build doesn't understand is skipped
/// rather than taking the whole payload down with it. The failure mode we want is
/// always "nothing to show", never "app is broken".
struct RemoteConfig: Equatable {
    var flags: Flags
    var messages: [RemoteMessage]

    static let empty = RemoteConfig(flags: Flags(), messages: [])
}

// MARK: - Flags

/// Behaviour the app reads silently. Only `minSupportedVersion` draws any UI.
struct Flags: Equatable {
    var minSupportedVersion = "0.0.0"
    var updateURL: URL?
    var features = Features()
}

/// Kill switches. All on by default — turning one off is the only thing that has
/// any effect, which keeps an unreachable worker from costing anyone a feature.
struct Features: Equatable {
    var pip = true
    var appleSignIn = true
    var googleSignIn = true
}

// MARK: - Messages

/// One thing to show, on one surface.
struct RemoteMessage: Identifiable, Equatable {
    /// Where the message is rendered. A surface this build doesn't know about
    /// means the message can't be drawn at all, so it's dropped in decoding.
    enum Surface: String, Decodable {
        case homeBanner
        case settingsRow
    }

    enum Severity: String, Decodable {
        case info
        case warning
        case critical
    }

    struct Action: Equatable {
        enum Kind: String, Decodable {
            case openURL
            case appStore
            case dismiss
        }

        var kind: Kind
        var label: String
        var url: URL?
    }

    /// Narrows the audience. The worker filters on this too; the client re-checks
    /// because a cached payload can outlive the app version it was fetched for.
    struct Match: Equatable {
        var platforms: [String]?
        var minVersion: String?
        var maxVersion: String?
        var minBuild: Int?
        var locales: [String]?
    }

    var id: String
    var surface: Surface
    var severity: Severity = .info
    var priority = 0
    var title: String
    var body: String?
    var actions: [Action] = []
    var dismissible = true
    var expiresAt: Date?
    var rolloutPercent = 100
    var match: Match?

    /// Hosts a message is allowed to link to, mirrored from the worker. A link
    /// anywhere else means someone got at the response, so the message is dropped
    /// rather than shown without its action.
    static let allowedHosts: Set<String> = [
        "cuecard.dev", "www.cuecard.dev", "apps.apple.com", "github.com"
    ]

    /// Whether this is safe and sensible to draw: a title, no more than two
    /// actions, every link pointing somewhere we recognise, and some way out.
    var isRenderable: Bool {
        guard !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard actions.count <= 2 else { return false }

        for action in actions where action.kind == .openURL {
            guard let host = action.url?.host, Self.allowedHosts.contains(host) else { return false }
        }

        return dismissible || !actions.isEmpty
    }

    func hasExpired(asOf now: Date = Date()) -> Bool {
        guard let expiresAt else { return false }
        return expiresAt <= now
    }
}

// MARK: - Targeting

extension RemoteMessage.Match {
    func admits(platform: String, version: String, build: Int, locale: String) -> Bool {
        if let platforms, !platforms.contains(platform) { return false }
        if let minVersion, AppVersion.compare(version, minVersion) == .orderedAscending { return false }
        if let maxVersion, AppVersion.compare(version, maxVersion) == .orderedDescending { return false }
        if let minBuild, build < minBuild { return false }

        // Language alone: "en" covers en-US, en-GB and the rest.
        if let locales {
            let language = locale.split(separator: "-").first.map { $0.lowercased() } ?? locale.lowercased()
            let admitted = locales.contains { $0.split(separator: "-").first?.lowercased() == language }
            if !admitted { return false }
        }

        return true
    }
}

// MARK: - Version numbers

enum AppVersion {
    static var current: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    }

    static var build: Int {
        Int(Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "") ?? 0
    }

    /// Compare two dotted versions numerically. Missing components count as zero,
    /// so "1.2" and "1.2.0" are the same version.
    static func compare(_ lhs: String, _ rhs: String) -> ComparisonResult {
        let left = lhs.split(separator: ".").map { Int($0) ?? 0 }
        let right = rhs.split(separator: ".").map { Int($0) ?? 0 }

        for index in 0..<max(left.count, right.count) {
            let l = index < left.count ? left[index] : 0
            let r = index < right.count ? right[index] : 0
            if l != r { return l < r ? .orderedAscending : .orderedDescending }
        }

        return .orderedSame
    }
}

// MARK: - Decoding

/// Decodes an element, or nothing at all. Lets one malformed or not-yet-understood
/// message drop out without discarding the messages around it.
private struct Skipping<Wrapped: Decodable>: Decodable {
    let value: Wrapped?

    init(from decoder: Decoder) throws {
        value = try? Wrapped(from: decoder)
    }
}

/// The worker writes plain ISO 8601; accept fractional seconds too in case that
/// ever changes underneath us.
private func parseISO8601(_ string: String) -> Date? {
    let withFraction = ISO8601DateFormatter()
    withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

    return ISO8601DateFormatter().date(from: string) ?? withFraction.date(from: string)
}

extension RemoteConfig: Decodable {
    private enum CodingKeys: String, CodingKey {
        case flags, messages
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        flags = (try? container.decode(Flags.self, forKey: .flags)) ?? Flags()

        let decoded = (try? container.decode([Skipping<RemoteMessage>].self, forKey: .messages)) ?? []
        messages = decoded.compactMap(\.value)
    }
}

extension Flags: Decodable {
    private enum CodingKeys: String, CodingKey {
        case minSupportedVersion, updateURL, features
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        minSupportedVersion = (try? container.decode(String.self, forKey: .minSupportedVersion)) ?? "0.0.0"
        updateURL = try? container.decode(URL.self, forKey: .updateURL)
        features = (try? container.decode(Features.self, forKey: .features)) ?? Features()
    }
}

extension Features: Decodable {
    private enum CodingKeys: String, CodingKey {
        case pip, appleSignIn, googleSignIn
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        pip = (try? container.decode(Bool.self, forKey: .pip)) ?? true
        appleSignIn = (try? container.decode(Bool.self, forKey: .appleSignIn)) ?? true
        googleSignIn = (try? container.decode(Bool.self, forKey: .googleSignIn)) ?? true
    }
}

extension RemoteMessage: Decodable {
    private enum CodingKeys: String, CodingKey {
        case id, surface, severity, priority, title, body
        case actions, dismissible, expiresAt, rolloutPercent, match
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        // An id, a surface and a title are the message. Without all three there's
        // nothing to draw or remember, so let the decode fail and skip it.
        id = try container.decode(String.self, forKey: .id)
        surface = try container.decode(Surface.self, forKey: .surface)
        title = try container.decode(String.self, forKey: .title)

        severity = (try? container.decode(Severity.self, forKey: .severity)) ?? .info
        priority = (try? container.decode(Int.self, forKey: .priority)) ?? 0
        body = try? container.decode(String.self, forKey: .body)
        dismissible = (try? container.decode(Bool.self, forKey: .dismissible)) ?? true
        rolloutPercent = (try? container.decode(Int.self, forKey: .rolloutPercent)) ?? 100
        match = try? container.decode(Match.self, forKey: .match)

        if let expiry = try? container.decode(String.self, forKey: .expiresAt) {
            expiresAt = parseISO8601(expiry)
        }

        // An action kind we don't recognise is one we can't carry out, and the
        // action is usually the point of the message — so drop the message rather
        // than show a card whose button does nothing.
        actions = try container.decodeIfPresent([Action].self, forKey: .actions) ?? []
    }
}

extension RemoteMessage.Action: Decodable {
    private enum CodingKeys: String, CodingKey {
        case kind, label, url
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        kind = try container.decode(Kind.self, forKey: .kind)
        label = try container.decode(String.self, forKey: .label)
        url = try? container.decode(URL.self, forKey: .url)
    }
}

extension RemoteMessage.Match: Decodable {
    private enum CodingKeys: String, CodingKey {
        case platforms, minVersion, maxVersion, minBuild, locales
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        platforms = try? container.decode([String].self, forKey: .platforms)
        minVersion = try? container.decode(String.self, forKey: .minVersion)
        maxVersion = try? container.decode(String.self, forKey: .maxVersion)
        minBuild = try? container.decode(Int.self, forKey: .minBuild)
        locales = try? container.decode([String].self, forKey: .locales)
    }
}

File: cuecard-mobile/android/app/src/main/java/com/thisisnsh/cuecard/android/Models/RemoteConfig.kt
package com.thisisnsh.cuecard.android.models

import com.thisisnsh.cuecard.android.BuildConfig
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import java.net.URI
import java.time.Instant
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

/**
 * The payload behind `GET /v2/config` on the mobile worker.
 *
 * Decoding is deliberately forgiving. Every flag defaults to its permissive
 * value, so a build that can't reach the worker behaves exactly as shipped, and
 * anything a newer worker adds that this build doesn't understand is skipped
 * rather than taking the whole payload down with it. The failure mode we want is
 * always "nothing to show", never "app is broken".
 */
data class RemoteConfig(
    val flags: Flags = Flags(),
    val messages: List<RemoteMessage> = emptyList()
) {
    companion object {
        val EMPTY = RemoteConfig()

        /** Lenient by design: see the class comment. */
        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        /**
         * Decode a worker response. One unreadable message drops out without
         * taking the rest of the payload with it — the parallel to the iOS
         * `Skipping<Wrapped>` wrapper.
         */
        fun decode(raw: String): RemoteConfig? {
            val root = runCatching { json.parseToJsonElement(raw).jsonObject }.getOrNull() ?: return null

            val flags = (root["flags"] as? JsonObject)
                ?.let { runCatching { json.decodeFromJsonElement(FlagsDTO.serializer(), it) }.getOrNull() }
                ?.toFlags()
                ?: Flags()

            val messages = (root["messages"] as? JsonArray).orEmpty().mapNotNull { element ->
                runCatching { json.decodeFromJsonElement(RemoteMessageDTO.serializer(), element) }
                    .getOrNull()
                    ?.toMessage()
            }

            return RemoteConfig(flags = flags, messages = messages)
        }
    }
}

// MARK: - Flags

/** Behaviour the app reads silently. Only `minSupportedVersion` draws any UI. */
data class Flags(
    val minSupportedVersion: String = "0.0.0",
    val updateURL: String? = null,
    val features: Features = Features()
)

/**
 * Kill switches. All on by default — turning one off is the only thing that has
 * any effect, which keeps an unreachable worker from costing anyone a feature.
 */
data class Features(
    val pip: Boolean = true,
    val appleSignIn: Boolean = true,
    val googleSignIn: Boolean = true
)

// MARK: - Messages

/** One thing to show, on one surface. */
data class RemoteMessage(
    val id: String,
    val surface: Surface,
    val severity: Severity = Severity.INFO,
    val priority: Int = 0,
    val title: String,
    val body: String? = null,
    val actions: List<Action> = emptyList(),
    val dismissible: Boolean = true,
    val expiresAt: Instant? = null,
    val rolloutPercent: Int = 100,
    val match: Match? = null
) {
    /**
     * Where the message is rendered. A surface this build doesn't know about
     * means the message can't be drawn at all, so it's dropped in decoding.
     */
    enum class Surface(val rawValue: String) {
        HOME_BANNER("homeBanner"),
        SETTINGS_ROW("settingsRow");

        companion object {
            fun from(value: String?): Surface? = entries.find { it.rawValue == value }
        }
    }

    enum class Severity(val rawValue: String) {
        INFO("info"),
        WARNING("warning"),
        CRITICAL("critical");

        companion object {
            fun from(value: String?): Severity? = entries.find { it.rawValue == value }
        }
    }

    data class Action(
        val kind: Kind,
        val label: String,
        val url: String? = null
    ) {
        enum class Kind(val rawValue: String) {
            OPEN_URL("openURL"),
            APP_STORE("appStore"),
            DISMISS("dismiss");

            companion object {
                fun from(value: String?): Kind? = entries.find { it.rawValue == value }
            }
        }
    }

    /**
     * Narrows the audience. The worker filters on this too; the client re-checks
     * because a cached payload can outlive the app version it was fetched for.
     */
    data class Match(
        val platforms: List<String>? = null,
        val minVersion: String? = null,
        val maxVersion: String? = null,
        val minBuild: Int? = null,
        val locales: List<String>? = null
    ) {
        fun admits(platform: String, version: String, build: Int, locale: String): Boolean {
            if (platforms != null && !platforms.contains(platform)) return false
            if (minVersion != null && AppVersion.compare(version, minVersion) < 0) return false
            if (maxVersion != null && AppVersion.compare(version, maxVersion) > 0) return false
            if (minBuild != null && build < minBuild) return false

            // Language alone: "en" covers en-US, en-GB and the rest.
            if (locales != null) {
                val language = locale.split("-").firstOrNull()?.lowercase() ?: locale.lowercase()
                val admitted = locales.any { it.split("-").firstOrNull()?.lowercase() == language }
                if (!admitted) return false
            }

            return true
        }
    }

    /**
     * Whether this is safe and sensible to draw: a title, no more than two
     * actions, every link pointing somewhere we recognise, and some way out.
     */
    val isRenderable: Boolean
        get() {
            if (title.trim().isEmpty()) return false
            if (actions.size > 2) return false

            for (action in actions.filter { it.kind == Action.Kind.OPEN_URL }) {
                val url = action.url ?: return false
                val host = runCatching { URI(url).host }.getOrNull() ?: return false
                if (!allowedHosts.contains(host)) return false
            }

            return dismissible || actions.isNotEmpty()
        }

    fun hasExpired(now: Instant = Instant.now()): Boolean {
        val expiry = expiresAt ?: return false
        return !expiry.isAfter(now)
    }

    companion object {
        /**
         * Hosts a message is allowed to link to, mirrored from the worker. A link
         * anywhere else means someone got at the response, so the message is
         * dropped rather than shown without its action.
         */
        val allowedHosts = setOf(
            "cuecard.dev", "www.cuecard.dev", "play.google.com", "github.com"
        )
    }
}

// MARK: - Version numbers

object AppVersion {
    val current: String
        get() = BuildConfig.VERSION_NAME

    val build: Int
        get() = BuildConfig.VERSION_CODE

    /**
     * Compare two dotted versions numerically. Missing components count as zero,
     * so "1.2" and "1.2.0" are the same version.
     */
    fun compare(lhs: String, rhs: String): Int {
        val left = lhs.split(".").map { it.toIntOrNull() ?: 0 }
        val right = rhs.split(".").map { it.toIntOrNull() ?: 0 }

        for (index in 0 until maxOf(left.size, right.size)) {
            val l = left.getOrElse(index) { 0 }
            val r = right.getOrElse(index) { 0 }
            if (l != r) return if (l < r) -1 else 1
        }

        return 0
    }
}

// MARK: - Decoding

/**
 * The worker writes plain ISO 8601; accept fractional seconds too in case that
 * ever changes underneath us.
 */
private fun parseISO8601(value: String): Instant? =
    runCatching { OffsetDateTime.parse(value, DateTimeFormatter.ISO_OFFSET_DATE_TIME).toInstant() }
        .recoverCatching { Instant.parse(value) }
        .getOrNull()

@Serializable
private data class FeaturesDTO(
    val pip: Boolean? = null,
    val appleSignIn: Boolean? = null,
    val googleSignIn: Boolean? = null
) {
    fun toFeatures() = Features(
        pip = pip ?: true,
        appleSignIn = appleSignIn ?: true,
        googleSignIn = googleSignIn ?: true
    )
}

@Serializable
private data class FlagsDTO(
    val minSupportedVersion: String? = null,
    @SerialName("updateURL") val updateUrl: String? = null,
    val features: FeaturesDTO? = null
) {
    fun toFlags() = Flags(
        minSupportedVersion = minSupportedVersion ?: "0.0.0",
        updateURL = updateUrl,
        features = features?.toFeatures() ?: Features()
    )
}

@Serializable
private data class ActionDTO(
    val kind: String,
    val label: String,
    val url: String? = null
) {
    /**
     * An action kind we don't recognise is one we can't carry out, and the action
     * is usually the point of the message — so this throws, and the message goes
     * with it.
     */
    fun toAction(): RemoteMessage.Action {
        val parsed = RemoteMessage.Action.Kind.from(kind)
            ?: throw IllegalArgumentException("Unknown action kind: $kind")
        return RemoteMessage.Action(kind = parsed, label = label, url = url)
    }
}

@Serializable
private data class MatchDTO(
    val platforms: List<String>? = null,
    val minVersion: String? = null,
    val maxVersion: String? = null,
    val minBuild: Int? = null,
    val locales: List<String>? = null
) {
    fun toMatch() = RemoteMessage.Match(
        platforms = platforms,
        minVersion = minVersion,
        maxVersion = maxVersion,
        minBuild = minBuild,
        locales = locales
    )
}

@Serializable
private data class RemoteMessageDTO(
    val id: String,
    val surface: String,
    val severity: String? = null,
    val priority: Int? = null,
    val title: String,
    val body: String? = null,
    val actions: List<ActionDTO>? = null,
    val dismissible: Boolean? = null,
    val expiresAt: String? = null,
    val rolloutPercent: Int? = null,
    val match: MatchDTO? = null
) {
    /**
     * An id, a surface and a title are the message. Without all three there's
     * nothing to draw or remember, so this throws and the message is skipped.
     */
    fun toMessage(): RemoteMessage {
        val parsedSurface = RemoteMessage.Surface.from(surface)
            ?: throw IllegalArgumentException("Unknown surface: $surface")

        return RemoteMessage(
            id = id,
            surface = parsedSurface,
            severity = RemoteMessage.Severity.from(severity) ?: RemoteMessage.Severity.INFO,
            priority = priority ?: 0,
            title = title,
            body = body,
            actions = actions?.map { it.toAction() } ?: emptyList(),
            dismissible = dismissible ?: true,
            expiresAt = expiresAt?.let { parseISO8601(it) },
            rolloutPercent = rolloutPercent ?: 100,
            match = match?.toMatch()
        )
    }
}
